// sqlectron-term/lib/containers/app.js
'use strict';

var _get = require('babel-runtime/helpers/get')['default'];

var _inherits = require('babel-runtime/helpers/inherits')['default'];

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _toConsumableArray = require('babel-runtime/helpers/to-consumable-array')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _widgetsShortcuts = require('../widgets/shortcuts');

var _widgetsShortcuts2 = _interopRequireDefault(_widgetsShortcuts);

var _widgetsStatus = require('../widgets/status');

var _widgetsStatus2 = _interopRequireDefault(_widgetsStatus);

var App = (function (_Component) {
  _inherits(App, _Component);

  function App() {
    _classCallCheck(this, App);

    _get(Object.getPrototypeOf(App.prototype), 'constructor', this).apply(this, arguments);
  }

  _createClass(App, [{
    key: 'render',
    value: function render() {
      var _props = this.props;
      var children = _props.children;
      var status = _props.status;
      var theme = this.context.theme;

      var shortcuts = [{ key: 'q', label: 'Quit' }, { key: 'escape', label: 'Back' }].concat(_toConsumableArray(this.props.shortcuts));
      return _react2['default'].createElement(
        'box',
        { position: { top: 0, left: 0, bottom: 0, right: 0 }, style: theme.header },
        _react2['default'].createElement('text', { style: theme.header, content: ' SQLECTRON' }),
        _react2['default'].createElement(
          'box',
          { position: { top: 1, left: 0, right: 0, bottom: 2 }, style: theme.main },
          children
        ),
        _react2['default'].createElement(_widgetsStatus2['default'], { status: status }),
        _react2['default'].createElement(_widgetsShortcuts2['default'], { shortcuts: shortcuts })
      );
    }
  }], [{
    key: 'propTypes',
    value: {
      children: _react.PropTypes.node,
      status: _react.PropTypes.string,
      shortcuts: _react.PropTypes.array
    },
    enumerable: true
  }, {
    key: 'contextTypes',
    value: {
      theme: _react.PropTypes.object.isRequired
    },
    enumerable: true
  }]);

  return App;
})(_react.Component);

exports['default'] = (0, _reactRedux.connect)(function (state) {
  return {
    status: state.status,
    shortcuts: state.shortcuts
  };
})(App);
module.exports = exports['default'];
