// sqlectron-term/lib/containers/shortcuts.js
'use strict';

var _get = require('babel-runtime/helpers/get')['default'];

var _inherits = require('babel-runtime/helpers/inherits')['default'];

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _react = require('react');

var _reactRedux = require('react-redux');

var _actionsShortcuts = require('../actions/shortcuts');

var Shortcuts = (function (_Component) {
  _inherits(Shortcuts, _Component);

  function Shortcuts() {
    _classCallCheck(this, Shortcuts);

    _get(Object.getPrototypeOf(Shortcuts.prototype), 'constructor', this).apply(this, arguments);
  }

  _createClass(Shortcuts, [{
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      if (this.focused) this.props.dispatch((0, _actionsShortcuts.removeShortcuts)(this.props.items));
      this.focused = false;
    }
  }, {
    key: 'handleFocus',
    value: function handleFocus() {
      this.focused = true;
      this.props.dispatch((0, _actionsShortcuts.addShortcuts)(this.props.items));
    }
  }, {
    key: 'handleBlur',
    value: function handleBlur() {
      this.focused = false;
      this.props.dispatch((0, _actionsShortcuts.removeShortcuts)(this.props.items));
    }
  }, {
    key: 'handleKeypress',
    value: function handleKeypress(ch, info) {
      var shortcut = this.props.items.find(function (item) {
        return item.key === info.full;
      });
      if (!shortcut) return false;

      if (shortcut.handler) return shortcut.handler(shortcut);
      if (this.props.onShortcut) return this.props.onShortcut(shortcut);
    }
  }, {
    key: 'render',
    value: function render() {
      return (0, _react.cloneElement)(this.props.children, {
        onFocus: this.handleFocus.bind(this),
        onBlur: this.handleBlur.bind(this),
        onKeypress: this.handleKeypress.bind(this)
      });
    }
  }], [{
    key: 'propTypes',
    value: {
      dispatch: _react.PropTypes.func.isRequired,

      children: _react.PropTypes.element.isRequired,
      items: _react.PropTypes.array.isRequired,

      onShortcut: _react.PropTypes.func
    },
    enumerable: true
  }]);

  return Shortcuts;
})(_react.Component);

exports['default'] = (0, _reactRedux.connect)()(Shortcuts);
module.exports = exports['default'];
