// sqlectron-term/lib/reducers/connections.js
'use strict';

var _extends = require('babel-runtime/helpers/extends')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _actionsConnections = require('../actions/connections');

var _actionsServers = require('../actions/servers');

var initialState = {
  didInvalidate: true
};

exports['default'] = function (state, action) {
  if (state === undefined) state = initialState;

  switch (action.type) {
    case _actionsServers.SAVE_SERVER_SUCCESS:
      {
        if (state.server && action.server.id === state.server.id) {
          return _extends({}, state, { didInvalidate: true });
        }
        return state;
      }
    case _actionsConnections.DB_CONNECT_REQUEST:
      {
        var server = action.server;
        var database = action.database;

        return {
          isConnecting: true,
          isConnected: false,
          didInvalidate: false,
          server: server,
          database: database
        };
      }
    case _actionsConnections.DB_CONNECT_SUCCESS:
      {
        return _extends({}, state, { isConnecting: false, isConnected: true, error: null });
      }
    case _actionsConnections.DB_CONNECT_FAILURE:
      {
        return _extends({}, state, {
          isConnecting: false,
          error: action.error
        });
      }
    default:
      return state;
  }
};

module.exports = exports['default'];
