// sqlectron-term/lib/widgets/database-list.js
'use strict';

var _get = require('babel-runtime/helpers/get')['default'];

var _inherits = require('babel-runtime/helpers/inherits')['default'];

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _lodash = require('lodash');

var DatabaseList = (function (_Component) {
  _inherits(DatabaseList, _Component);

  function DatabaseList() {
    _classCallCheck(this, DatabaseList);

    _get(Object.getPrototypeOf(DatabaseList.prototype), 'constructor', this).apply(this, arguments);
  }

  _createClass(DatabaseList, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      if (this.refs.list) this.refs.list.focus();
    }
  }, {
    key: 'selected',
    value: function selected() {
      if (this.refs.list) return this.refs.list.selected;
      return 0;
    }
  }, {
    key: 'focus',
    value: function focus() {
      this.refs.list.focus();
    }
  }, {
    key: 'handleKeypress',
    value: function handleKeypress(ch, info) {
      if (this.props.onKeypress) this.props.onKeypress(ch, info);
    }
  }, {
    key: 'handleFocus',
    value: function handleFocus() {
      if (this.props.onFocus) this.props.onFocus();
    }
  }, {
    key: 'handleBlur',
    value: function handleBlur() {
      if (this.props.onBlur) this.props.onBlur();
    }
  }, {
    key: 'render',
    value: function render() {
      var items = this.props.items;
      var theme = this.context.theme;

      return _react2['default'].createElement('list', {
        keys: true, mouse: true, shadow: true, scrollbar: true,
        ref: 'list',
        style: (0, _lodash.merge)({}, theme.list.normal, theme.list.focus),
        position: { left: 'center', top: 'center' },
        border: 'line',
        label: ' Database list ',
        items: items,
        onFocus: this.handleFocus.bind(this),
        onBlur: this.handleBlur.bind(this),
        onKeypress: this.handleKeypress.bind(this)
      });
    }
  }], [{
    key: 'propTypes',
    value: {
      items: _react.PropTypes.array.isRequired,
      // events
      onFocus: _react.PropTypes.func,
      onBlur: _react.PropTypes.func,
      onKeypress: _react.PropTypes.func
    },
    enumerable: true
  }, {
    key: 'contextTypes',
    value: {
      theme: _react.PropTypes.object.isRequired
    },
    enumerable: true
  }]);

  return DatabaseList;
})(_react.Component);

exports['default'] = DatabaseList;
module.exports = exports['default'];
