// sqlectron-term/lib/default-theme.js
'use strict';

var _extends = require('babel-runtime/helpers/extends')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});
var red = '#ff0000';
var white = '#ffffff';
var blue = '#0000d3';
var green = '#00cfd0';
var cyan = '#00ffff';
var black = '#000000';
var yellow = '#ffff00';
var lightGray = '#e7e7e7';

var styles = {
  button: {
    fg: black,
    bg: lightGray,
    focus: {
      bg: green,
      fg: black
    }
  },
  header: { fg: yellow, bg: blue, bold: true },
  main: { bg: lightGray },
  status: { bg: green, fg: black },
  shortcuts: {
    general: {
      fg: yellow,
      bg: blue,
      bold: true
    },
    key: cyan
  },
  box: {
    normal: {
      bg: blue,
      fg: yellow,
      border: { fg: white, bg: blue },
      scrollbar: { bg: green },
      label: { fg: white, bg: blue }
    },
    focus: {
      border: { fg: cyan }
    }
  },
  danger: {
    bg: red,
    border: { fg: white, bg: red }
  }
};

styles.list = _extends({}, styles.box);
styles.list.normal.item = { fg: yellow };

exports['default'] = styles;
module.exports = exports['default'];
